README.md
# Puppet lab projects

This purpose of this project is to install and configure puppet and learn the basic usage of puppet, and how automation goes in puppet.

The labs will consist of three nodes with one master, one worker and one node. it will be run in a virtualbox hypervisor with vagrant.
all the commands will be controlled via code. That means that we will use infrastructure as a code wherever it is possible.

## Default configuration of puppet server
This default configuration assigned to the puppet server settings for this labs.

```sh
sudo vi /etc/puppetlabs/puppet/puppet.conf
```

Place the following lines, Modify them to match your environment.

```sh
[master]
dns_alt_names = master.puppetvincelabs.com,master

[main]
certname = master.puppetvincelabs.com
server = master.puppetvincelabs.com
environment = production
runinterval = 15m
```

## Default configuration of puppet agent
This default configuration assigned to the puppet agent settings for this labs.

```sh
sudo vi /etc/puppetlabs/puppet/puppet.conf
```

Place the following lines, Modify them to match your environment.

```sh

[main]
certname = client.puppetvincelabs.com
server = master.puppetvincelabs.com
environment = production
runinterval = 15m
```


CHANGELOG.md
# Change Log

## [unreleased]

## 2020-07-28
### Added

* Creation of the main files  
* Structuring the project.
* Documentation of the main idea for the project.

## 2020-07-30
### Added

* adding the bootstrap scripts for agent and master.
* Testing the process to create the environment.
* Adjusting the vagrantfile by removing unnecessary codes.
* redesigning the .gitignore.
* Documentation of the puppet lab.


scripts/bootstrap-master.sh
#!/bin/sh
#title           : bootstrap-master.sh
#description     : This script will bootstrap installation and configuration of puppet master nodes on ubuntu.
#author		     : vincenira
#date            : 2020729
#version         : 0.1    
#usage		     : sh bootstrap-master.sh
#notes           : Install Vim and Emacs to use this script.
#bash_version    : 0.0.1-unreleased
#==============================================================================

# Run on VM to bootstrap Puppet Master Server
export DEBIAN_FRONTEND=noninteractive
installer_ntp()
{
    sudo apt-get update -yq && sudo apt-get install -yq ntp ntpdate
    sudo ntpdate -u 0.ubuntu.pool.ntp.org
    date && timedatectl list-timezones && sudo timedatectl set-timezone America/New_York
}

if ps aux | grep "puppet master" | grep -v grep 2> /dev/null; then
    
    echo "Puppet Master is already installed. Existing"
    exit 0
else
    # Install NTP client
    installer_ntp
    # Install Puppet Master
    wget https://apt.puppetlabs.com/puppet6-release-bionic.deb
    sudo dpkg -i puppet6-release-bionic.deb
    sudo DEBIAN_FRONTEND=noninteractive apt-get update -yq 
    sudo DEBIAN_FRONTEND=noninteractive apt-get upgrade -yq  
    sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq puppetserver
    # Add optional alternate DNS names to /etc/puppetlabs/puppet/puppet.conf
    if [ -e "/etc/puppetlabs/puppet/puppet.conf" ]; then
        if grep -Fxq "\[main\]" /etc/puppetlabs/puppet/puppet.conf 
        then
            sudo sed -i 's/.*\[master\].*/&\ndns_alt_names = master,master.puppetvincelabs.com/' /etc/puppetlabs/puppet/puppet.conf
            sudo sed -i 's/.*\[master\].*/&\nautosign = true/' /etc/puppetlabs/puppet/puppet.conf
        else
            sudo sed -i 's/.*\[master\].*/&\ndns_alt_names = master,master.puppetvincelabs.com/' /etc/puppetlabs/puppet/puppet.conf
            sudo sed -i 's/.*\[master\].*/&\nautosign = true/' /etc/puppetlabs/puppet/puppet.conf
            echo "[main]" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
            echo "certname = $(hostname -f)" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
            echo "server = master.puppetvincelabs.com" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
            echo "environment = production" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null
            echo "runinterval = 15m" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null
        fi 
    fi
    sudo systemctl enable puppetserver && sudo systemctl start puppetserver
fi

scripts/bootstrap-agent.sh
#!/bin/sh
#title           : bootstrap-agent.sh
#description     : This script will bootstrap installation and configuration of puppet agent nodes on ubuntu.
#author		     : vincenira
#date            : 2020729
#version         : 0.1    
#usage		     : sh bootstrap-agent.sh
#notes           : Install Vim and Emacs to use this script.
#bash_version    : 0.0.1-unreleased
#==============================================================================
export DEBIAN_FRONTEND=noninteractive
config_etc()
{
        echo "" | sudo tee --append /etc/hosts 2> /dev/null && \
        echo "# Host config for Puppet Master and Agent Nodes" | sudo tee --append /etc/hosts 2> /dev/null && \
        echo "192.168.56.15   master.puppetvincelabs.com  master" | sudo tee --append /etc/hosts 2> /dev/null && \
        echo "192.168.56.25   node01.puppetvincelabs.com  node01" | sudo tee --append /etc/hosts 2> /dev/null && \
        echo "192.168.56.35   node02.puppetvincelabs.com  node02" | sudo tee --append /etc/hosts 2> /dev/null
}

config_puppet_conf()
{
        echo "" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
        echo "[main]" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
        echo "certname = $(hostname -f)" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
        echo "server = master.puppetvincelabs.com" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null && \
        echo "environment = production" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null
        echo "runinterval = 15m" | sudo tee --append /etc/puppetlabs/puppet/puppet.conf 2> /dev/null
}

enable_puppet()
{
    sudo /opt/puppetlabs/bin/puppet resource service puppet ensure=running enable=true
}

installer_git()
{

    sudo DEBIAN_FRONTEND=noninteractive apt-get update -yq 
    sudo DEBIAN_FRONTEND=noninteractive apt-get upgrade -yq
    sudo DEBIAN_FRONTEND=noninteractive apt-get install git -yq
}

installer_puppet_agent()
{
    wget https://apt.puppetlabs.com/puppet6-release-bionic.deb
    sudo dpkg -i puppet6-release-bionic.deb
    sudo DEBIAN_FRONTEND=noninteractive apt-get update -yq 
    sudo DEBIAN_FRONTEND=noninteractive apt-get upgrade -yq
    sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq puppet-agent
}

installer_ntp()
{
    sudo DEBIAN_FRONTEND=noninteractive apt-get update -yq 
    sudo DEBIAN_FRONTEND=noninteractive apt-get install -yq ntp ntpdate
    sudo ntpdate -u 0.ubuntu.pool.ntp.org
    date && timedatectl list-timezones && sudo timedatectl set-timezone America/New_York
}
# Run on  VM to bootstrap Puppet Agent nodes

if ps aux | grep "puppet agent" | grep -v grep 2> /dev/null; then
    echo "Puppet Agent is already installed. Moving on..."
    # Configure /etc/hosts file
    if cat /etc/hosts | grep "puppetvincelabs.com" 2> /dev/null; then
        exit 0
    else
        config_etc
        enable_puppet
        installer_git
        exit 0
    fi 
else
    installer_ntp
    installer_puppet_agent
    installer_git
    config_etc
    config_puppet_conf
    enable_puppet
fi

# if cat /etc/crontab | grep puppet 2> /dev/null; then
#     echo "Puppet Agent is already configured. Existing..."
#     exit 0
# else
#     # Add a cron from puppet to run every 30 mins.
#         sudo puppet resource cron puppet-agent ensure=present user=root minute 30 \
#         command='/usr/bin/puppet agent --onetime --no-daemonize --splay'

#         sudo puppet resource service puppet ensure=running enable=true
#         # Configure /etc/hosts file
#         config_etc
#         # Add agent section to /etc/puppet/puppet.conf
#         echo "" && echo "[agent]\nserver=master" | sudo tee --append /etc/puppet/puppet.conf 2> /dev/null
#         sudo puppet agent --enable
# fi
# # Enable puppet and give puppet time to auto sign the certificate and kick off a run.
#     sudo puppet agent --enable
#     sudo puppet agent -t --waitforcert 15

# added the code below as puppet runs produced a non-zero exist
exit 0